src/backend/cuda/jit.cpp

```diff
@@ -155,21 +179,31 @@ struct Param
         // Generate input offsets, only needs current id
         node->genOffsets(offsetsStream, ids_curr.id, is_linear);
         // Generate the core function body, needs children ids as well
-        node->genFuncs(opsStream, ids_curr);
+        node->genFuncs(opsStream, ids_curr, is_linear);
     }
 
-    outrefstream << "const Param<" << full_nodes[output_ids[0]]->getTypeStr()
-                 << "> &outref = out" << output_ids[0] << ";\n";
-
     for (int i = 0; i < (int)output_ids.size(); i++) {
         int id = output_ids[i];
         // Generate output parameters
-        outParamStream << "Param<" << full_nodes[id]->getTypeStr() << "> out"
-                       << id << ", \n";
+        outParamStream << "Param out" << id << ", \n";
+
         // Generate code to write the output
-        outWriteStream << "out" << id << ".ptr[idx] = val" << id << ";\n";
+        outWriteStream << "((" << full_nodes[id]->getTypeStr() << "*)(out" << id
+                       << ".ptr))"
+                       << "[idx] = val" << id << ";\n";
     }
 
+    // TODO(umar): find the ideal size for the offset array. This
+    // size shouldn't be too large I would think. I am sure someone
+    // will find a way to hit this bug waiting to happen.
+    paramreadstream << R"JIT(
+        extern __shared__ Param params[];
+        long long offsets[8];
```

Comment:
We should be able to do the following and avoid the hard coded declaration of offsets array. Assuming, we already included "Shared.hpp" thats available via buildKernel function.
```
SharedMemory<unsigned char> shared;
unsigned char* shrdMem = shared.getPointer();

Param* params = (Param*)shrdMem;
long long* offsets = (long long*)(&shrdMem[offsets_offset]);
```
`offsets_offset` could be passed along with `num_params`. Once we have `offsets` pointer, we would only access offsets array for only `num_params` values.

Reply:
offset is a local array not a shared array

Reply:
May be I wasn't clear, I am asking you to make it shared along with `Param`s. Making it a fixed size `8` and then waiting for the bug to happen(as described by you in the TODO comment) is not correct.

Reply:
After our little chat, I do see that offsets are based on global thread ids, hence have to be stored for each thread in the block. We can still do it by setting the shared memory size to the following
```
params.size() * (sizeof(Param<T>) + sizeof(long long)*total_num_of_threads_in_a_block)
```